feat(OGC-68): M1 - Backend Core (Storage Location CRUD + Device Connectivity)

pmanko opened the pull request:
## Summary

- Add storage location CRUD backend: deletion validation for Room/Device/Shelf/Rack with referential integrity checks
- Add device connectivity fields (ipAddress, port, communicationProtocol) with Liquibase schema migration
- Add FHIR R4 extensions for device connectivity in Location resource
- Add RBAC enforcement on DELETE endpoints (admin-only) with 403 response for non-admins
- Add name uniqueness validation within parent scope (FR-004)

## Related Issues

- Jira: [OGC-68](https://uwdigi.atlassian.net/browse/OGC-68)
- Spec PR: #2425 (merged)

## Changes

| Category | Files | Description |
|----------|-------|-------------|
| Controller | `StorageLocationRestController.java` | DELETE endpoints with RBAC + validation, device connectivity in response |
| Service | `StorageLocationService*.java`, `DeletionValidationResult.java` | Deletion validation, name uniqueness |
| DAO | `Storage*DAO*.java` | findByName methods for uniqueness checks |
| Entity | `StorageDevice.java`, `StorageDeviceForm.java` | ipAddress, port, communicationProtocol fields |
| FHIR | `StorageLocationFhirTransform.java` | Device connectivity extensions |
| Schema | `storage-device-connectivity.xml` | Liquibase changeset for new columns |
| Tests | `StorageLocationServiceTest.java`, `StorageLocationRestControllerTest.java` | Unit + integration tests |

## Test Plan

- [x] Unit tests pass: `mvn test -Dtest=StorageLocationServiceTest` (13 tests)
- [x] Integration tests pass: `mvn test -Dtest=StorageLocationRestControllerTest` (8 tests)
- [x] Build passes: `mvn clean install -DskipTests -Dmaven.test.skip=true`
- [x] Spotless formatting applied

## Constitution Compliance

- [x] Layered architecture (Principle IV)
- [x] FHIR R4 compliance (Principle III)
- [x] Liquibase for schema changes (Principle VI)
- [x] Test-driven development (Principle V)

[OGC-68]: https://uwdigi.atlassian.net/browse/OGC-68?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

pmanko committed:
feat(OGC-68): Add storage location CRUD backend and device connectivity

Implements Milestone 1 (Backend) of OGC-68 Storage Location CRUD feature.

Changes:
- Add connectivity fields (ipAddress, port, communicationProtocol) to StorageDevice entity
- Add Liquibase changeset for new device connectivity columns with port range validation
- Create DeletionValidationResult class for deletion validation responses
- Implement deletion validation methods (canDeleteRoom, canDeleteDevice, canDeleteShelf, canDeleteRack)
- Add DELETE endpoints for all storage location types (rooms, devices, shelves, racks)
- Update StorageDeviceForm DTO with connectivity fields and validation
- Add FHIR extensions for device connectivity (device-ip-address, device-port, device-communication-protocol)
- Add unit tests for service layer deletion validation (9 tests passing)
- Add integration tests for controller DELETE endpoints

Test Coverage:
- Service layer unit tests: 9/9 passing
- All tests use TDD Red-Green-Refactor approach

Related: OGC-68, Jira ticket https://uwdigi.atlassian.net/browse/OGC-68

pmanko committed:
fix(OGC-68): Add RBAC enforcement and name uniqueness validation

- Add admin check to DELETE endpoints (security fix)
- Remove duplicate DELETE endpoint mappings
- Add name uniqueness validation for locations within parent scope
- Add DAO findByName methods for uniqueness checks
- Add RBAC controller tests (T008a)
- Add unit tests for name uniqueness validation (T017a)
- Include connectivity fields in device response entityToMap

pmanko committed:
chore: Remove dev artifact references from comments

- Remove OGC-68 ticket references from code comments
- Remove T0xx task ID references from test files
- Replace with meaningful descriptive comments

mozzy11 commented on src/main/java/org/openelisglobal/storage/controller/StorageLocationRestController.java:
```diff
@@ -902,24 +936,18 @@ public ResponseEntity<Map<String, Object>> deleteRack(@PathVariable String id, H
             }
 
             boolean isAdmin = checkAdminStatus(request);
```
we would actually need to update this method [isUserAdmin](https://github.com/DIGI-UW/OpenELIS-Global-2/blob/b2496689a4185f25750369b3f50f9728c10d7a54/src/main/java/org/openelisglobal/login/daoimpl/UserModuleServiceImpl.java#L269) to simply check for GlobalAdmin role rather than the property [isAdmin](https://github.com/DIGI-UW/OpenELIS-Global-2/blob/b2496689a4185f25750369b3f50f9728c10d7a54/src/main/java/org/openelisglobal/login/valueholder/LoginUser.java#L73) because the property cant be assinged via the UI , but rather programatically

pmanko replied:
see https://uwdigi.atlassian.net/browse/OGC-172

mozzy11 commented on src/main/java/org/openelisglobal/storage/controller/StorageLocationRestController.java:
```diff
@@ -1156,6 +1184,9 @@ private Map<String, Object> entityToMap(Object entity) {
             map.put("capacityLimit", device.getCapacityLimit());
             map.put("active", device.getActive());
             map.put("fhirUuid", device.getFhirUuidAsString());
+            map.put("ipAddress", device.getIpAddress());
```
This is fine but in general its better to use strongly typed objects rather than loosely typed maps for response objects

mozzy11 commented on src/test/java/org/openelisglobal/storage/controller/StorageLocationRestControllerTest.java:
```diff
@@ -0,0 +1,197 @@
+    @Test
+    public void testDeleteRoom_WithValidId_Returns204() throws Exception {
+        // Arrange - Create room with no children
+        jdbcTemplate.update(
```
we prefer using dbUnit xml datasets for cleaner and more easiliy mantainable Tests , but we can refactor later

pmanko committed:
fix(OGC-68): address PR review feedback

Align storage deletion RBAC to Global Administrator role, improve device connectivity validation/migration defaults, and harden tests via DBUnit datasets and additional coverage.

pmanko commented:
## PR Review Feedback - Addressed

Thank you for the thorough review! All feedback has been addressed in the latest commit. Summary of changes:

### ✅ Admin RBAC (mozzy11)
**Feedback:** Update  to check GlobalAdmin role instead of  property.

**Resolution:** Implemented feature-specific RBAC in  using . This aligns storage deletion RBAC with the Global Administrator role without modifying core  (which is used across the codebase and kept unchanged for this PR).

**Note:** A broader authentication consistency issue has been identified and will be tracked in a separate Jira backlog issue for future core refactoring.

### ✅ Strongly Typed Response DTOs (mozzy11)
**Feedback:** Prefer strongly typed objects over loosely typed Maps for response objects.

**Resolution:** Introduced typed response DTOs (, , , , ) and updated single-entity endpoints to return these DTOs. JSON structure preserved for frontend compatibility.

### ✅ DBUnit XML Datasets (mozzy11)
**Feedback:** Prefer DBUnit XML datasets for cleaner, more maintainable tests.

**Resolution:** Refactored  to use  dataset, removing inline SQL setup/teardown.

### ✅ IPv6 Validation (Copilot)
**Feedback:** IPv6 regex doesn't handle compressed notation (e.g., `::1`) or IPv4-mapped addresses.

**Resolution:** Updated `StorageDeviceForm.ipAddress` regex to support:
- Compressed IPv6 notation (`::1`, `2001:db8::1`)
- IPv4-mapped IPv6 (`::ffff:192.0.2.1`)
- All valid IPv6 group counts (1-8 groups)
- Added comprehensive validation tests in `StorageDeviceFormTest`

### ✅ Liquibase Default Value (Copilot)
**Feedback:** `communication_protocol` default value "BACnet" will be misleading for existing devices.

**Resolution:** Removed `defaultValue="BACnet"` from the Liquibase changeset. Protocol is now only set explicitly when creating network-connected devices.

### ✅ Test Coverage Improvements (Copilot)
**Feedback:** Missing positive test cases for name uniqueness validation and RBAC tests for Device/Shelf/Rack deletion.

**Resolution:**
- Added positive test cases: `testIsNameUniqueWithinParent_RoomUnique_WhenDaoReturnsNull()` and `testIsNameUniqueWithinParent_DeviceUnique_WhenExcludeIdMatchesExisting()`
- Added 403 RBAC tests: `testDeleteDevice_AsNonAdminUser_Returns403()`, `testDeleteShelf_AsNonAdminUser_Returns403()`, `testDeleteRack_AsNonAdminUser_Returns403()`

### ✅ Code Quality (Copilot)
**Feedback:** Missing `@Override` annotation on `setUp()` method.

**Resolution:** Added `@Override` annotation to `StorageLocationRestControllerTest.setUp()`.

---

All changes have been tested, formatted, and committed. Ready for re-review!

pmanko commented:
## PR Review Feedback - Addressed

Thank you for the thorough review! All feedback has been addressed in the latest commit. Summary of changes:

### ✅ Admin RBAC (mozzy11)
**Feedback:** Update `isUserAdmin()` to check GlobalAdmin role instead of `isAdmin` property.

**Resolution:** Implemented feature-specific RBAC in `StorageLocationRestController.checkAdminStatus()` using `UserRoleService.userInRole(sysUserId, Constants.ROLE_GLOBAL_ADMIN)`. This aligns storage deletion RBAC with the Global Administrator role without modifying core `UserModuleServiceImpl` (which is used across the codebase and kept unchanged for this PR).

**Note:** A broader authentication consistency issue has been identified and will be tracked in a separate Jira backlog issue for future core refactoring.

### ✅ Strongly Typed Response DTOs (mozzy11)
**Feedback:** Prefer strongly typed objects over loosely typed Maps for response objects.

**Resolution:** Introduced typed response DTOs (`StorageRoomResponse`, `StorageDeviceResponse`, `StorageShelfResponse`, `StorageRackResponse`, `StorageBoxResponse`) and updated single-entity endpoints to return these DTOs. JSON structure preserved for frontend compatibility.

### ✅ DBUnit XML Datasets (mozzy11)
**Feedback:** Prefer DBUnit XML datasets for cleaner, more maintainable tests.

**Resolution:** Refactored `StorageLocationRestControllerTest` to use `storage-location-controller.xml` dataset, removing inline SQL setup/teardown.

### ✅ IPv6 Validation (Copilot)
**Feedback:** IPv6 regex doesn't handle compressed notation (e.g., `::1`) or IPv4-mapped addresses.

**Resolution:** Updated `StorageDeviceForm.ipAddress` regex to support:
- Compressed IPv6 notation (`::1`, `2001:db8::1`)
- IPv4-mapped IPv6 (`::ffff:192.0.2.1`)
- All valid IPv6 group counts (1-8 groups)
- Added comprehensive validation tests in `StorageDeviceFormTest`

### ✅ Liquibase Default Value (Copilot)
**Feedback:** `communication_protocol` default value "BACnet" will be misleading for existing devices.

**Resolution:** Removed `defaultValue="BACnet"` from the Liquibase changeset. Protocol is now only set explicitly when creating network-connected devices.

### ✅ Test Coverage Improvements (Copilot)
**Feedback:** Missing positive test cases for name uniqueness validation and RBAC tests for Device/Shelf/Rack deletion.

**Resolution:**
- Added positive test cases: `testIsNameUniqueWithinParent_RoomUnique_WhenDaoReturnsNull()` and `testIsNameUniqueWithinParent_DeviceUnique_WhenExcludeIdMatchesExisting()`
- Added 403 RBAC tests: `testDeleteDevice_AsNonAdminUser_Returns403()`, `testDeleteShelf_AsNonAdminUser_Returns403()`, `testDeleteRack_AsNonAdminUser_Returns403()`

### ✅ Code Quality (Copilot)
**Feedback:** Missing `@Override` annotation on `setUp()` method.

**Resolution:** Added `@Override` annotation to `StorageLocationRestControllerTest.setUp()`.

---

All changes have been tested, formatted, and committed. Ready for re-review!

pmanko committed:
fix(OGC-68): refactor cascade delete test to use DBUnit XML datasets

Replace inline SQL inserts with DBUnit XML dataset to follow established test data management patterns and prevent test data pollution.

- Create storage-cascade-delete-test.xml DBUnit dataset
- Remove inline SQL createTestHierarchy() method
- Remove manual cleanup methods (DBUnit handles state management)
- Update tests to use constants from dataset instead of hardcoded IDs

pmanko committed:
test(docs): clarify DBUnit datasets and fix cascade test wiring

Make DBUnit dataset usage explicit in developer docs and align examples with repository conventions. Fix StorageLocationRestControllerCascadeDeleteTest to construct JdbcTemplate from DataSource (no JdbcTemplate bean).

pmanko commented:
Coordination note:

- M2 frontend work is tracked in **#2432**, which is currently **stacked on this branch**.
- Once CI is green here and this merges to `develop`, #2432 can merge cleanly (or be retargeted to `develop` if needed).

(Closed duplicate combined PR #2440 to keep milestone PRs clean.)

pmanko committed:
fix(e2e): reset to /login between authentication cases

The login authentication loop could leave a user logged in, so a page reload would not show the login form and the next iteration would fail to find #loginName. Visiting /login each time and logging out after successful login stabilizes the spec.

pmanko commented:
CI fix:

- The failing `build-and-run-qa-tests` check was due to Cypress `login.cy.js` (`Validates user authentication`) leaving a user logged in during the loop, so subsequent iterations could not find `#loginName`.
- Pushed fix commit: `60585eb71` (visit `/login` each iteration + sign out after successful login).

This should stabilize the login E2E spec in CI.

pmanko committed:
chore: apply Spotless formatting

pmanko committed:
docs(testing): align roadmap with repo conventions

pmanko committed:
test(fixtures): remove storage hierarchy duplication

pmanko committed:
Fix storage DBUnit fixtures to match schema

pmanko committed:
Stop seeding storage test data via Liquibase

pmanko committed:
Add DBUnit fixture loader for E2E/manual testing

- Create DbUnitFixtureLoader.java: standalone loader for DBUnit XML datasets
- Update load-test-fixtures.sh: use DBUnit loader instead of SQL+Liquibase assumption
- Storage fixtures now use same source-of-truth as JUnit tests
- Removes dependency on Liquibase test context for storage hierarchy

pmanko committed:
Format code with spotless

pmanko committed:
Make DBUnit loader port configurable via DB_PORT env var

- Port defaults to 15432 (matches all docker-compose files)
- Can be overridden via DB_PORT environment variable
- Consistent across dev/test/build/main environments

pmanko committed:
Update port comment for clarity

pmanko committed:
fix(e2e): fix login.cy test with boolean fix + API logout

- Change Users.json to use booleans (not strings "true"/"false")
- Fix validation loop: if (user.correctPass) instead of === true
- Add cy.ensureLoggedOut() API-based logout after successful login
- Remove flaky menuButton check (mainHeader sufficient for auth verification)

This eliminates the session leak that caused #loginName not found errors
when the 2nd successful login in the validation loop stayed authenticated.
